csharp/ql/lib/semmle/code/csharp/commons/QualifiedName.qll

```diff
@@ -1,6 +1,201 @@
+/** Provides predicates for working with fully qualified names. */
+
+private import csharp
+private import dotnet
+
 /**
- * Provides predicates related to C# qualified names.
+ * Holds if namespace `n` has the qualified name `qualifier`.`name`.
+ *
+ * For example if the qualified name is `System.Collections.Generic`, then
+ * `qualifier`=`System.Collections` and `name`=`Generic`.
  */
+predicate namespaceHasQualifiedName(DotNet::Namespace n, string qualifier, string name) {
+  if n instanceof DotNet::GlobalNamespace
+  then qualifier = "" and name = ""
+  else (
+    exists(string pqualifier, string pname |
+      namespaceHasQualifiedName(n.getParentNamespace(), pqualifier, pname) and
+      qualifier = getQualifiedName(pqualifier, pname)
+    ) and
+    name = n.getName()
+  )
+}
+
+/** Gets a string of `N` commas where `N + 1` is the number of type parameters of this unbound generic. */
+private string getTypeParameterCommas(UnboundGeneric ug) {
+  result = strictconcat(int i | exists(ug.getTypeParameter(i)) | "", ",")
+}
+
+/** Provides the input to `QualifiedName`. */
+signature module QualifiedNameInputSig {
+  /** Gets the suffix to print after unbound generic `ug`. */
+  default string getUnboundGenericSuffix(UnboundGeneric ug) {
+    result = "<" + getTypeParameterCommas(ug) + ">"
+  }
+}
+
+/** Provides predicates for computing fully qualified names. */
+module QualifiedName<QualifiedNameInputSig Input> {
+  private string getDimensionString(ArrayType at, Type elementType) {
+    exists(Type et, string res |
+      et = at.getElementType() and
+      res = at.getArraySuffix() and
+      if et instanceof ArrayType
+      then result = res + getDimensionString(et, elementType)
+      else (
+        result = res and elementType = et
+      )
+    )
+  }
+
+  bindingset[t]
+  private string getFullName(Type t) {
+    exists(string qualifier, string name |
+      hasQualifiedName(t, qualifier, name) and
+      result = getQualifiedName(qualifier, name)
+    )
+  }
+
+  /** Gets the concatenation of the `getFullName` of type arguments. */
+  language[monotonicAggregates]
+  private string getTypeArgumentsQualifiedNames(ConstructedGeneric cg) {
+    result =
+      strictconcat(Type t, int i | t = cg.getTypeArgument(i) | getFullName(t), "," order by i)
+  }
+
+  /** Holds if declaration `d` has the qualified name `qualifier`.`name`. */
+  predicate hasQualifiedName(Declaration d, string qualifier, string name) {
+    d =
+      any(ValueOrRefType vort |
+        vort instanceof VoidType and
+        qualifier = "System" and
+        name = "Void"
```

Comment:
Why do we need to special case `VoidType`? 
Couldn't the below cover it?
```
          not exists(vort.getDeclaringType()) and
          qualifier = vort.getNamespace().getFullName() and
          name = vort.getUndecoratedName()
```

Reply:
Perhaps; I just copied over the existing logic.